MSH/msh_mesh.cpp

```diff
@@ -4583,5 +3575,102 @@ void CFEMesh::GetConnectedElements(std::vector<long>& nodes_on_sfc, std::vector<
 		}
 	}
 }
+
+/*!
+   Find element nodes on the top surface of a mesh domain
+   07.06.2010
+   By WW
+ */
+void CFEMesh::markTopSurfaceFaceElements3D()
+{
+#ifdef output_top_z
+	/// For output z coordinate of all nodes on the top surface
+	/// 13.08.2010. WW
+	vector<bool> node_mark(NodesNumber_Linear);
+	for (std::size_t i = 0; i < NodesNumber_Linear; i++)
+		node_mark[i] = false;
+#endif
+
+	const double tol = sqrt(std::numeric_limits<double>::epsilon()); // 1.e-5;
+	for (std::size_t i = 0; i < face_vector.size(); i++)
+	{
+		CElem* elem = face_vector[i];
+		CElem const* const own_elem = elem->GetOwner();
+
+		//// In element
+		//		// compute center of mass
+		double cent[] = {0., 0., 0.};
+		for (std::size_t k = 0; k < own_elem->GetNodesNumber(false); k++)
+		{
+			//				node = own_elem->nodes[k];
+			double const* const pnt_k(own_elem->GetNode(k)->getData());
+			cent[0] += pnt_k[0];
+			cent[1] += pnt_k[1];
+			cent[2] += pnt_k[2];
+		}
+		for (int k = 0; k < 3; k++)
+			cent[k] /= (double)own_elem->GetNodesNumber(false);
+
+		//			node = elem->nodes[0];
+		double const* const pnt_0(elem->GetNode(0)->getData());
+		cent[0] -= pnt_0[0];
+		cent[1] -= pnt_0[1];
+		cent[2] -= pnt_0[2];
+		NormalizeVector(cent, 3);
+		elem->ComputeVolume();
+		elem->FillTransformMatrix();
+		/// Compute the normal to this surface element
+		Math_Group::Matrix const* const transform_tensor = elem->getTransformTensor();
+		double fac = cent[0] * (*transform_tensor)(0, 2) + cent[1] * (*transform_tensor)(1, 2)
+		                   + cent[2] * (*transform_tensor)(2, 2);
+		if (fac > 0.0)
+			fac = -1.0;
+		else
+			fac = 1.0;
+		//////
+
+		/// If n.z>0
+		if ((*transform_tensor)(2, 2) * fac > tol)
+		{
+			elem->SetMark(true);
+			for (int k = 0; k < 3; k++)
+				(*transform_tensor)(k, 2) *= fac;
+
+#ifdef output_top_z
+			for (std::size_t k = 0; k < elem->GetNodesNumber(false); k++)
+				node_mark[elem->GetNodeIndex(k)] = true;
+#endif
+		}
+		else if ((*transform_tensor)(2, 2) * fac < -tol)
+		{
+			elem->SetMark(false);
+			for (int k = 0; k < 3; k++)
+				(*transform_tensor)(k, 2) *= fac;
+
+#ifdef output_top_z
+			for (k = 0; k < elem->GetNodesNumber(quad); k++)
+				node_mark[elem->GetNodeIndex(k)] = bottom;
+#endif
+		}
+		else
+			elem->SetMark(false);
```

Comment:
are you sure if the element marks are set true at initial?

Reply:
Here the face elements on the top surface are marked by the condition of
n * z >0, where n is the normal vector of an face element, and z=(0, 0, 1)^T. The orientation of the normal vector is guaranteed by a vector from the the center of 3D element to an node of the face.

In this loop over face elements, elements that satisfy n * z >0 are marked as true, and the others are marked as false. Therefore there is no need to set initial status of the face elements.

Reply:
In the new commit, I moved the ``else if`` for the bottom surface to the section of `` #ifdef output_top_z``.